File: game/cave/scylla_lost.py
import pygame.mixer

from animation import Animation
from cave.cave_resources import CaveResources
from cave.family_cage_opens import FamilyCageOpens
from cave.scylla_spring import ScyllaSpring
from game_data import GameData
from phone_events import PhoneEvents
from state import State


class ScyllaLost(State):
    hugo_sprite_pos = [
        (25, 105),
        (97, 100),
        (172, 102),
    ]

    def __init__(self, context: GameData):
        super().__init__(context)
        self.animation = [CaveResources.scylla_bird, CaveResources.scylla_leaves, CaveResources.scylla_ropes][context.cave_win_type]

    def process_events(self, phone_events: PhoneEvents):
        if self.get_frame_index() >= len(self.animation):
            return ScyllaSpring if self.context.cave_win_type == 2 else FamilyCageOpens

        if self.context.cave_win_type == 0: # Bird
            if self.one_shot(0.5, "FugleSkrig"):
                pygame.mixer.Sound.play(CaveResources.fugle_skrig)
        elif self.context.cave_win_type == 1: # Leaves
            if self.one_shot(0.5, "Skrig"):
                pygame.mixer.Sound.play(CaveResources.skrig)
        else: # Ropes
            if self.one_shot(1, "PrePuf"):
                pygame.mixer.Sound.play(CaveResources.pre_puf)
            if self.one_shot(2, "Puf"):
                pygame.mixer.Sound.play(CaveResources.puf)
        return None


    def render(self, screen):
        screen.blit(Animation.get_frame(self.animation, self.get_frame_index()), (0,0))
        screen.blit(CaveResources.hugo[0], self.hugo_sprite_pos[self.context.cave_selected_rope])

File: game-web/src/cave/scyllaLost.ts
/** Port of game/cave/scylla_lost.py */

import { State, type StateClass } from '../state';
import { Animation } from '../animation';
import { AudioHelper } from '../audioHelper';
import { CaveResources } from './caveResources';
import { FamilyCageOpens } from './familyCageOpens';
import { ScyllaSpring } from './scyllaSpring';
import type { GameData } from '../gameData';
import type { PhoneEvents } from '../phoneEvents';

export class ScyllaLost extends State {
  private static readonly hugoSpritePos: [number, number][] = [
    [25, 105],
    [97, 100],
    [172, 102],
  ];

  private animation: ImageBitmap[];

  constructor(context: GameData) {
    super(context);
    this.animation = [
      CaveResources.scyllaBird,
      CaveResources.scyllaLeaves,
      CaveResources.scyllaRopes,
    ][context.caveWinType];
  }

  processEvents(_events: PhoneEvents): StateClass | null {
    if (this.getFrameIndex() >= this.animation.length) {
      return this.context.caveWinType === 2 ? ScyllaSpring : FamilyCageOpens;
    }

    if (this.context.caveWinType === 0) {
      if (this.oneShot(0.5, 'FugleSkrig')) AudioHelper.play(CaveResources.fugleSkrig);
    } else if (this.context.caveWinType === 1) {
      if (this.oneShot(0.5, 'Skrig')) AudioHelper.play(CaveResources.skrig);
    } else {
      if (this.oneShot(1, 'PrePuf')) AudioHelper.play(CaveResources.prePuf);
      if (this.oneShot(2, 'Puf'))    AudioHelper.play(CaveResources.puf);
    }

    return null;
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    ctx.drawImage(Animation.getFrame(this.animation, this.getFrameIndex()), 0, 0);
    const [hx, hy] = ScyllaLost.hugoSpritePos[this.context.caveSelectedRope];
    ctx.drawImage(CaveResources.hugo[0], hx, hy);
  }
}
